src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py

```diff
@@ -0,0 +1,475 @@
+        """
+        self._batteries = batteries
+
+    @classmethod
+    @abstractmethod
+    def name(cls) -> str:
+        """Return name of the formula.
+
+        Returns:
+            Name of the formula
+        """
+
+    @property
+    @abstractmethod
+    def battery_metrics(self) -> Dict[int, List[ComponentMetricId]]:
+        """Return what metrics are needed for each battery.
+
+        Returns:
+            Map between battery id and set of required metrics id.
+        """
+
+    @property
+    @abstractmethod
+    def inverter_metrics(self) -> Dict[int, List[ComponentMetricId]]:
+        """Return what metrics are needed for each inverter.
+
+        Returns:
+            Map between inverter id and set of required metrics id.
+        """
+
+    @abstractmethod
+    def calculate(
+        self,
+        metrics_data: Dict[int, ComponentMetricsData],
+        working_batteries: Set[int],
+    ) -> Optional[T]:
+        """Aggregate the metrics_data and calculate high level metric.
+
+        Missing components will be ignored. Formula will be calculated for all
+        working batteries that are in metrics_data.
+
+        Args:
+            metrics_data: Components metrics data, that should be used to calculate the
+                result.
+            working_batteries: working batteries. These batteries will be used
+                to calculate the result. It should be subset of the batteries given in a
+                constructor.
+
+        Returns:
+            High level metric calculated from the given metrics.
+            Return None if there are no component metrics.
+        """
+
+    def _is_missing(
+        self, component_id: int, components_metrics: Dict[int, ComponentMetricsData]
+    ) -> bool:
+        if component_id not in components_metrics:
+            _logger.warning(
+                "%s: Missing metrics from the components %d",
+                self.name,
+                component_id,
+            )
+            return True
+
+        return False
+
+
+# Soc and capacity are different kind of metrics.
+# It makes sens to separate them with `_`.
+class SoC_Capacity_Calculator(
+    MetricCalculator[SoC_Capacity]
+):  # pylint: disable=invalid-name
+    """Define how to calculate SoC and Capacity metrics."""
+
+    def __init__(self, batteries: Set[int]) -> None:
+        """Create class instance.
+
+        Args:
+            batteries: What batteries should be used for calculation.
+        """
+
+        super().__init__(batteries)
+
+        self._battery_metrics = [
+            ComponentMetricId.CAPACITY,
+            ComponentMetricId.SOC_LOWER_BOUND,
+            ComponentMetricId.SOC_UPPER_BOUND,
+            ComponentMetricId.SOC,
+        ]
+
+    @classmethod
+    def name(cls) -> str:
+        """Return name of the calculator.
+
+        Returns:
+            Name of the calculator
+        """
+        return "Soc_Capacity"
```

Comment:
If this is only for debugging, maybe you can move it to `MetricCalculator` and use an implementation like `return cls.__name__`.

Reply:
No, I use is outside in battery_pool module to create unique name for the active methods

Reply:
Done, thanks!